File: fraiseql-go/fraiseql/observers_test.go
package fraiseql

import (
	"encoding/json"
	"testing"
)

func TestObserverBuilder(t *testing.T) {
	// Clear registry
	Reset()

	// Build observer
	NewObserver("onOrderCreated").
		Entity("Order").
		Event("INSERT").
		Actions(Webhook("https://example.com/orders")).
		Register()

	schema := GetSchema()

	if len(schema.Observers) != 1 {
		t.Fatalf("Expected 1 observer, got %d", len(schema.Observers))
	}

	observer := schema.Observers[0]
	if observer.Name != "onOrderCreated" {
		t.Errorf("Expected name 'onOrderCreated', got '%s'", observer.Name)
	}
	if observer.Entity != "Order" {
		t.Errorf("Expected entity 'Order', got '%s'", observer.Entity)
	}
	if observer.Event != "INSERT" {
		t.Errorf("Expected event 'INSERT', got '%s'", observer.Event)
	}
	if len(observer.Actions) != 1 {
		t.Fatalf("Expected 1 action, got %d", len(observer.Actions))
	}
	if observer.Actions[0]["type"] != "webhook" {
		t.Errorf("Expected action type 'webhook', got '%s'", observer.Actions[0]["type"])
	}
}

func TestObserverWithCondition(t *testing.T) {
	Reset()

	NewObserver("onHighValueOrder").
		Entity("Order").
		Event("INSERT").
		Condition("total > 1000").
		Actions(Webhook("https://example.com")).
		Register()

	schema := GetSchema()
	observer := schema.Observers[0]

	if observer.Condition != "total > 1000" {
		t.Errorf("Expected condition 'total > 1000', got '%s'", observer.Condition)
	}
}

func TestObserverWithCustomRetry(t *testing.T) {
	Reset()

	customRetry := RetryConfig{
		MaxAttempts:     5,
		BackoffStrategy: "linear",
		InitialDelayMs:  200,
		MaxDelayMs:      30000,
	}

	NewObserver("onOrder").
		Entity("Order").
		Event("INSERT").
		Actions(Webhook("https://example.com")).
		Retry(customRetry).
		Register()

	schema := GetSchema()
	observer := schema.Observers[0]

	if observer.Retry.MaxAttempts != 5 {
		t.Errorf("Expected max_attempts 5, got %d", observer.Retry.MaxAttempts)
	}
	if observer.Retry.BackoffStrategy != "linear" {
		t.Errorf("Expected backoff_strategy 'linear', got '%s'", observer.Retry.BackoffStrategy)
	}
	if observer.Retry.InitialDelayMs != 200 {
		t.Errorf("Expected initial_delay_ms 200, got %d", observer.Retry.InitialDelayMs)
	}
	if observer.Retry.MaxDelayMs != 30000 {
		t.Errorf("Expected max_delay_ms 30000, got %d", observer.Retry.MaxDelayMs)
	}
}

func TestObserverWithDefaultRetry(t *testing.T) {
	Reset()

	NewObserver("onOrder").
		Entity("Order").
		Event("INSERT").
		Actions(Webhook("https://example.com")).
		Register()

	schema := GetSchema()
	observer := schema.Observers[0]

	if observer.Retry.MaxAttempts != DefaultRetryConfig.MaxAttempts {
		t.Errorf("Expected default max_attempts %d, got %d", DefaultRetryConfig.MaxAttempts, observer.Retry.MaxAttempts)
	}
	if observer.Retry.BackoffStrategy != DefaultRetryConfig.BackoffStrategy {
		t.Errorf("Expected default backoff_strategy '%s', got '%s'", DefaultRetryConfig.BackoffStrategy, observer.Retry.BackoffStrategy)
	}
}

func TestObserverWithMultipleActions(t *testing.T) {
	Reset()

	NewObserver("onOrder").
		Entity("Order").
		Event("INSERT").
		Actions(
			Webhook("https://example.com/orders"),
			Slack("#orders", "New order {id}"),
			EmailAction("admin@example.com", "Order created", "Order {id} created"),
		).
		Register()

	schema := GetSchema()
	observer := schema.Observers[0]

	if len(observer.Actions) != 3 {
		t.Fatalf("Expected 3 actions, got %d", len(observer.Actions))
	}
	if observer.Actions[0]["type"] != "webhook" {
		t.Errorf("Expected action 0 type 'webhook', got '%s'", observer.Actions[0]["type"])
	}
	if observer.Actions[1]["type"] != "slack" {
		t.Errorf("Expected action 1 type 'slack', got '%s'", observer.Actions[1]["type"])
	}
	if observer.Actions[2]["type"] != "email" {
		t.Errorf("Expected action 2 type 'email', got '%s'", observer.Actions[2]["type"])
	}
}

func TestWebhookAction(t *testing.T) {
	action := Webhook("https://example.com/orders")

	if action["type"] != "webhook" {
		t.Errorf("Expected type 'webhook', got '%s'", action["type"])
	}
	if action["url"] != "https://example.com/orders" {
		t.Errorf("Expected url 'https://example.com/orders', got '%s'", action["url"])
	}
	if headers, ok := action["headers"].(map[string]string); !ok {
		t.Error("Expected headers to be map[string]string")
	} else if headers["Content-Type"] != "application/json" {
		t.Errorf("Expected Content-Type 'application/json', got '%s'", headers["Content-Type"])
	}
}

func TestWebhookWithEnvVar(t *testing.T) {
	action := WebhookWithEnv("ORDER_WEBHOOK_URL")

	if action["type"] != "webhook" {
		t.Errorf("Expected type 'webhook', got '%s'", action["type"])
	}
	if action["url_env"] != "ORDER_WEBHOOK_URL" {
		t.Errorf("Expected url_env 'ORDER_WEBHOOK_URL', got '%s'", action["url_env"])
	}
	if _, exists := action["url"]; exists {
		t.Error("Expected url to not exist when using url_env")
	}
}

func TestWebhookWithCustomHeaders(t *testing.T) {
	action := Webhook("https://example.com", map[string]interface{}{
		"headers": map[string]string{
			"Authorization": "Bearer token123",
		},
	})

	if headers, ok := action["headers"].(map[string]string); !ok {
		t.Error("Expected headers to be map[string]string")
	} else if headers["Authorization"] != "Bearer token123" {
		t.Errorf("Expected Authorization 'Bearer token123', got '%s'", headers["Authorization"])
	}
}

func TestWebhookWithBodyTemplate(t *testing.T) {
	action := Webhook("https://example.com", map[string]interface{}{
		"body_template": `{"order_id": "{{id}}"}`,
	})

	if action["body_template"] != `{"order_id": "{{id}}"}` {
		t.Errorf("Expected body_template to match, got '%s'", action["body_template"])
	}
}

func TestSlackAction(t *testing.T) {
	action := Slack("#orders", "New order {id}: ${total}")

	if action["type"] != "slack" {
		t.Errorf("Expected type 'slack', got '%s'", action["type"])
	}
	if action["channel"] != "#orders" {
		t.Errorf("Expected channel '#orders', got '%s'", action["channel"])
	}
	if action["message"] != "New order {id}: ${total}" {
		t.Errorf("Expected message, got '%s'", action["message"])
	}
	if action["webhook_url_env"] != "SLACK_WEBHOOK_URL" {
		t.Errorf("Expected webhook_url_env 'SLACK_WEBHOOK_URL', got '%s'", action["webhook_url_env"])
	}
}

func TestSlackWithCustomWebhook(t *testing.T) {
	action := Slack("#orders", "New order", map[string]interface{}{
		"webhook_url": "https://hooks.slack.com/services/XXX",
	})

	if action["webhook_url"] != "https://hooks.slack.com/services/XXX" {
		t.Errorf("Expected webhook_url, got '%s'", action["webhook_url"])
	}
	if _, exists := action["webhook_url_env"]; exists {
		t.Error("Expected webhook_url_env to not exist when using webhook_url")
	}
}

func TestSlackWithCustomEnvVar(t *testing.T) {
	action := Slack("#alerts", "Alert!", map[string]interface{}{
		"webhook_url_env": "SLACK_ALERTS_WEBHOOK",
	})

	if action["webhook_url_env"] != "SLACK_ALERTS_WEBHOOK" {
		t.Errorf("Expected webhook_url_env 'SLACK_ALERTS_WEBHOOK', got '%s'", action["webhook_url_env"])
	}
}

func TestEmailAction(t *testing.T) {
	action := EmailAction("admin@example.com", "Order {id} created", "Order {id} for ${total} was created")

	if action["type"] != "email" {
		t.Errorf("Expected type 'email', got '%s'", action["type"])
	}
	if action["to"] != "admin@example.com" {
		t.Errorf("Expected to 'admin@example.com', got '%s'", action["to"])
	}
	if action["subject"] != "Order {id} created" {
		t.Errorf("Expected subject, got '%s'", action["subject"])
	}
	if action["body"] != "Order {id} for ${total} was created" {
		t.Errorf("Expected body, got '%s'", action["body"])
	}
}

func TestEmailWithFrom(t *testing.T) {
	action := EmailAction("customer@example.com", "Order shipped", "Your order is on its way!", map[string]interface{}{
		"from_email": "noreply@example.com",
	})

	if action["from"] != "noreply@example.com" {
		t.Errorf("Expected from 'noreply@example.com', got '%s'", action["from"])
	}
}

func TestSchemaExportWithObservers(t *testing.T) {
	Reset()

	NewObserver("onOrder1").
		Entity("Order").
		Event("INSERT").
		Actions(Webhook("https://example.com")).
		Register()

	NewObserver("onOrder2").
		Entity("Order").
		Event("UPDATE").
		Actions(Slack("#orders", "Updated")).
		Register()

	schema := GetSchema()

	if len(schema.Observers) != 2 {
		t.Fatalf("Expected 2 observers, got %d", len(schema.Observers))
	}
	if schema.Observers[0].Name != "onOrder1" && schema.Observers[1].Name != "onOrder1" {
		t.Error("Expected onOrder1 in observers")
	}
	if schema.Observers[0].Name != "onOrder2" && schema.Observers[1].Name != "onOrder2" {
		t.Error("Expected onOrder2 in observers")
	}
}

func TestSchemaJSONWithObservers(t *testing.T) {
	Reset()

	NewObserver("onOrder").
		Entity("Order").
		Event("INSERT").
		Actions(Webhook("https://example.com")).
		Register()

	schemaJSON, err := GetSchemaJSON(true)
	if err != nil {
		t.Fatalf("Failed to get schema JSON: %v", err)
	}

	var schema map[string]interface{}
	if err := json.Unmarshal(schemaJSON, &schema); err != nil {
		t.Fatalf("Failed to unmarshal schema JSON: %v", err)
	}

	if _, exists := schema["observers"]; !exists {
		t.Error("Expected observers key in schema JSON")
	}
}

func TestEmptyObservers(t *testing.T) {
	Reset()

	RegisterType("User", []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
	}, "User type")

	schema := GetSchema()

	// Observers should be empty array, not nil
	if schema.Observers == nil {
		// This is acceptable - Go will omit empty slices in JSON
		return
	}
	if len(schema.Observers) != 0 {
		t.Errorf("Expected 0 observers, got %d", len(schema.Observers))
	}
}

File: sdks/official/fraiseql-typescript/tests/observers.test.ts
/**
 * Tests for observer authoring API.
 */

import { describe, it, expect, beforeEach } from "@jest/globals";
import {
  Observer,
  webhook,
  slack,
  email,
  DEFAULT_RETRY_CONFIG,
  SchemaRegistry,
} from "../src/index";
import type { RetryConfig } from "../src/index";

beforeEach(() => {
  SchemaRegistry.clear();
});

describe("Observer decorator", () => {
  it("should register observer with schema registry", () => {
    class Observers {
      @Observer({
        entity: "Order",
        event: "INSERT",
        actions: [webhook("https://example.com")],
      })
      onOrderCreated() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();

    expect(schema.observers).toBeDefined();
    expect(schema.observers?.length).toBe(1);

    const observer = schema.observers![0];
    expect(observer.name).toBe("onOrderCreated");
    expect(observer.entity).toBe("Order");
    expect(observer.event).toBe("INSERT");
    expect(observer.actions.length).toBe(1);
    expect(observer.actions[0].type).toBe("webhook");
  });

  it("should support condition expressions", () => {
    class Observers {
      @Observer({
        entity: "Order",
        event: "UPDATE",
        condition: "status == 'paid'",
        actions: [webhook("https://example.com")],
      })
      onOrderPaid() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();
    const observer = schema.observers![0];

    expect(observer.condition).toBe("status == 'paid'");
  });

  it("should support custom retry configuration", () => {
    const retryConfig: RetryConfig = {
      max_attempts: 5,
      backoff_strategy: "linear",
      initial_delay_ms: 200,
      max_delay_ms: 30000,
    };

    class Observers {
      @Observer({
        entity: "Order",
        event: "INSERT",
        actions: [webhook("https://example.com")],
        retry: retryConfig,
      })
      onOrder() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();
    const observer = schema.observers![0];

    expect(observer.retry.max_attempts).toBe(5);
    expect(observer.retry.backoff_strategy).toBe("linear");
    expect(observer.retry.initial_delay_ms).toBe(200);
    expect(observer.retry.max_delay_ms).toBe(30000);
  });

  it("should use default retry config when not provided", () => {
    class Observers {
      @Observer({
        entity: "Order",
        event: "INSERT",
        actions: [webhook("https://example.com")],
      })
      onOrder() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();
    const observer = schema.observers![0];

    expect(observer.retry.max_attempts).toBe(DEFAULT_RETRY_CONFIG.max_attempts);
    expect(observer.retry.backoff_strategy).toBe(DEFAULT_RETRY_CONFIG.backoff_strategy);
  });

  it("should normalize event type to uppercase", () => {
    class Observers {
      @Observer({
        entity: "Order",
        event: "insert" as any, // Bypass type check for test
        actions: [webhook("https://example.com")],
      })
      onOrder() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();
    const observer = schema.observers![0];

    expect(observer.event).toBe("INSERT");
  });

  it("should support multiple actions", () => {
    class Observers {
      @Observer({
        entity: "Order",
        event: "INSERT",
        actions: [
          webhook("https://example.com/orders"),
          slack("#orders", "New order {id}"),
          email("admin@example.com", "Order created", "Order {id} created"),
        ],
      })
      onOrder() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();
    const observer = schema.observers![0];

    expect(observer.actions.length).toBe(3);
    expect(observer.actions[0].type).toBe("webhook");
    expect(observer.actions[1].type).toBe("slack");
    expect(observer.actions[2].type).toBe("email");
  });
});

describe("webhook action", () => {
  it("should create webhook with static URL", () => {
    const action = webhook("https://example.com/orders");

    expect(action.type).toBe("webhook");
    expect(action.url).toBe("https://example.com/orders");
    expect(action.headers).toEqual({ "Content-Type": "application/json" });
  });

  it("should create webhook with environment variable", () => {
    const action = webhook(undefined, { url_env: "ORDER_WEBHOOK_URL" });

    expect(action.type).toBe("webhook");
    expect(action.url_env).toBe("ORDER_WEBHOOK_URL");
    expect(action.url).toBeUndefined();
  });

  it("should support custom headers", () => {
    const action = webhook("https://example.com", {
      headers: { Authorization: "Bearer token123" },
    });

    expect(action.headers.Authorization).toBe("Bearer token123");
  });

  it("should support body template", () => {
    const action = webhook("https://example.com", {
      body_template: '{"order_id": "{{id}}"}',
    });

    expect(action.body_template).toBe('{"order_id": "{{id}}"}');
  });

  it("should throw error without URL or url_env", () => {
    expect(() => webhook()).toThrow("Either url or url_env must be provided");
  });
});

describe("slack action", () => {
  it("should create slack action with default webhook_url_env", () => {
    const action = slack("#orders", "New order {id}: ${total}");

    expect(action.type).toBe("slack");
    expect(action.channel).toBe("#orders");
    expect(action.message).toBe("New order {id}: ${total}");
    expect(action.webhook_url_env).toBe("SLACK_WEBHOOK_URL");
  });

  it("should support custom webhook URL", () => {
    const action = slack("#orders", "New order", {
      webhook_url: "https://hooks.slack.com/services/XXX",
    });

    expect(action.webhook_url).toBe("https://hooks.slack.com/services/XXX");
  });

  it("should support custom environment variable", () => {
    const action = slack("#alerts", "Alert!", {
      webhook_url_env: "SLACK_ALERTS_WEBHOOK",
    });

    expect(action.webhook_url_env).toBe("SLACK_ALERTS_WEBHOOK");
  });
});

describe("email action", () => {
  it("should create email action", () => {
    const action = email(
      "admin@example.com",
      "Order {id} created",
      "Order {id} for ${total} was created"
    );

    expect(action.type).toBe("email");
    expect(action.to).toBe("admin@example.com");
    expect(action.subject).toBe("Order {id} created");
    expect(action.body).toBe("Order {id} for ${total} was created");
  });

  it("should support custom sender", () => {
    const action = email(
      "customer@example.com",
      "Order shipped",
      "Your order is on its way!",
      { from_email: "noreply@example.com" }
    );

    expect(action.from).toBe("noreply@example.com");
  });
});

describe("Schema export", () => {
  it("should include observers in schema", () => {
    class Observers {
      @Observer({
        entity: "Order",
        event: "INSERT",
        actions: [webhook("https://example.com")],
      })
      onOrder1() {}

      @Observer({
        entity: "Order",
        event: "UPDATE",
        actions: [slack("#orders", "Updated")],
      })
      onOrder2() {}
    }
    void Observers;

    const schema = SchemaRegistry.getSchema();

    expect(schema.observers).toBeDefined();
    expect(schema.observers?.length).toBe(2);
    expect(schema.observers![0].name).toBe("onOrder1");
    expect(schema.observers![1].name).toBe("onOrder2");
  });

  it("should not include observers key when none defined", () => {
    const schema = SchemaRegistry.getSchema();

    expect(schema.observers).toBeUndefined();
  });
});
